### PR TITLE
enh(api/conf) add FindHostTemplates endpoint

### PR DESCRIPTION
**Pull request by a-launois**
## Description

Create new endpoint to list all host templates:
```
GET /configuration/hosts/templates
```

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x (master)
- [x] 23.10.x

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/src/Core/Common/Domain/BitmaskEnumTrait.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+namespace Core\Common\Domain;
+
+/**
+ * This trait is to be used for enums.
+ * It provides methods to transform data between enum to bitmask format.
+ * Example: 0b11 <=> 5 <=> [self::Down,self::Unreachable].
+ */
+trait BitmaskEnumTrait
+{
+    abstract public function toBit(): int;
```
Comment (abuathier) [inline centreon/src/Core/Common/Domain/BitmaskEnumTrait.php:33]:
```suggestion
    /**
     * Returns the int representation (2^n) of the bit of each enum.
     *
     * @return int
     */
    abstract public function toBit(): int;
```

File: centreon/src/Core/Common/Domain/BitmaskEnumTrait.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    abstract public static function getMaxBitmask(): int;
```
Comment (abuathier) [inline centreon/src/Core/Common/Domain/BitmaskEnumTrait.php:35]:
```suggestion
    /**
     * Returns the maximum possible bitmask as an int representation (2^0 + ... + 2^n).
     *
     * @return int
     */
    abstract public static function getMaxBitmask(): int;
```

File: centreon/src/Core/Common/Domain/BitmaskEnumTrait.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            // Value 0 is not a bit, we consider it resets the bitmask
+            if ($event->toBit() === 0) {
+                return 0;
+            }
+            $bitmask |= $event->toBit();
+        }
+
+        return $bitmask;
+    }
+}
```
Comment (abuathier) [inline centreon/src/Core/Common/Domain/BitmaskEnumTrait.php:80]: Code style : missing line feed at the end of the file

File: centreon/src/Core/Common/Domain/HostEvent.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            self::Recovery => 0b00100,
+            self::Flapping => 0b01000,
+            self::DowntimeScheduled => 0b10000,
+        };
+    }
+
+    public static function getMaxBitmask(): int {
+        return 0b11111;
+    }
+}
```
Comment (abuathier) [inline centreon/src/Core/Common/Domain/HostEvent.php:52]: Code style : missing line feed at the end of the file

File: centreon/src/Core/Common/Domain/HostEvent.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+
+declare(strict_types=1);
+
+namespace Core\Common\Domain;
+
+enum HostEvent: string
```
Comment (abuathier) [inline centreon/src/Core/Common/Domain/HostEvent.php:26]: Missing a brief phpdoc to explain the business meaning of this enum.

File: centreon/src/Core/Common/Domain/HostType.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+
+declare(strict_types=1);
+
+namespace Core\Common\Domain;
+
+enum HostType: string
```
Comment (abuathier) [inline centreon/src/Core/Common/Domain/HostType.php:26]: Missing a brief phpdoc to explain the business meaning of this enum.

File: centreon/src/Core/Common/Domain/SnmpVersion.php [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+
+declare(strict_types=1);
+
+namespace Core\Common\Domain;
+
+enum SnmpVersion: string
```
Comment (abuathier) [inline centreon/src/Core/Common/Domain/SnmpVersion.php:26]: Missing a brief phpdoc to explain the business meaning of this enum.

File: centreon/src/Core/Common/Domain/YesNoDefault.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+
+declare(strict_types=1);
+
+namespace Core\Common\Domain;
+
+enum YesNoDefault:string
```
Comment (abuathier) [inline centreon/src/Core/Common/Domain/YesNoDefault.php:26]: Missing a brief phpdoc to explain the business meaning of this enum.

File: centreon/src/Core/HostTemplate/Application/UseCase/FindHostTemplates/FindHostTemplates.php [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+use Core\Application\Common\UseCase\ForbiddenResponse;
+use Core\Application\Common\UseCase\PresenterInterface;
+use Core\Common\Domain\HostEvent;
+use Core\HostTemplate\Application\Exception\HostTemplateException;
+use Core\HostTemplate\Application\Repository\ReadHostTemplateRepositoryInterface;
+use Core\HostTemplate\Domain\Model\HostTemplate;
+use Core\HostTemplate\Infrastructure\API\FindHostTemplates\FindHostTemplatesPresenterOnPrem;
+use Core\HostTemplate\Infrastructure\API\FindHostTemplates\FindHostTemplatesPresenterSaas;
+
+final class FindHostTemplates
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Application/UseCase/FindHostTemplates/FindHostTemplates.php:40]: Plural everywhere ? `FindHostTemplates` VS `FindHostTemplate`

File: centreon/src/Core/HostTemplate/Application/UseCase/FindHostTemplates/FindHostTemplates.php [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+                'isPassiveCheckEnabled' => $hostTemplate->isPassiveCheckEnabled()->toInt(),
+                'isNotificationEnabled' => $hostTemplate->isNotificationEnabled()->toInt(),
+                /**
+                 * TODO
+                 *  this is related to api (> 23.10) behaviour, where no options is egal to a full bitmask
+                 *  Do we keep this behaviour or do we return null ?
+                 */
+                'notificationOptions' => $hostTemplate->getNotificationOptions() !== []
+                    ? HostEvent::toBitmask($hostTemplate->getNotificationOptions())
+                    : HostEvent::getMaxBitmask(),
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Application/UseCase/FindHostTemplates/FindHostTemplates.php:114]:
```suggestion
                'notificationOptions' => HostEvent::toBitmask(
                    $hostTemplate->getNotificationOptions() ?: HostEvent::getMaxBitmask()
                ),
```

File: centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+            $result['severity_id'],
+            $result['command_command_id'],
+            (string) $result['command_command_id_arg1'],
+            $result['timeperiod_tp_id'],
+            $result['host_max_check_attemps'],
+            $result['host_check_interval'],
+            $result['host_retry_check_interval'],
+            null !== $result['host_active_checks_enabled']
+                ? YesNoDefault::from($result['host_active_checks_enabled'])
+                : YesNoDefault::Default,
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php:267]:
```suggestion
            YesNoDefault::from($result['host_active_checks_enabled'] ?? YesNoDefault::Default),
```

File: centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+            null !== $result['host_passive_checks_enabled']
+                ? YesNoDefault::from($result['host_passive_checks_enabled'])
+                : YesNoDefault::Default,
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php:270]:
```suggestion
            YesNoDefault::from($result['host_passive_checks_enabled'] ?? YesNoDefault::Default),
```

File: centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+            null !== $result['host_notifications_enabled']
+                ? YesNoDefault::from($result['host_notifications_enabled'])
+                : YesNoDefault::Default,
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php:273]:
```suggestion
            YesNoDefault::from($result['host_notifications_enabled'] ?? YesNoDefault::Default),
```

File: centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+            $result['host_notification_interval'],
+            $result['timeperiod_tp_id2'],
+            (bool) $result['cg_additive_inheritance'],
+            (bool) $result['contact_additive_inheritance'],
+            $result['host_first_notification_delay'],
+            $result['host_recovery_notification_delay'],
+            $result['host_acknowledgement_timeout'],
+            null !== $result['host_check_freshness']
+                ? YesNoDefault::from($result['host_check_freshness'])
+                : YesNoDefault::Default,
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php:287]:
```suggestion
            YesNoDefault::from($result['host_check_freshness'] ?? YesNoDefault::Default),
```

File: centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+            $result['host_freshness_threshold'],
+            null !== $result['host_flap_detection_enabled']
+                ? YesNoDefault::from($result['host_flap_detection_enabled'])
+                : YesNoDefault::Default,
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php:291]:
```suggestion
            YesNoDefault::from($result['host_flap_detection_enabled'] ?? YesNoDefault::Default),
```

File: centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+            $result['host_low_flap_threshold'],
+            $result['host_high_flap_threshold'],
+            null !== $result['host_event_handler_enabled']
+                ? YesNoDefault::from($result['host_event_handler_enabled'])
+                : YesNoDefault::Default,
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Infrastructure/Repository/DbReadHostTemplateRepository.php:296]:
```suggestion
            YesNoDefault::from($result['host_event_handler_enabled'] ?? YesNoDefault::Default),
```

File: centreon/tests/php/Core/Common/Domain/EventEnumStub.php [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ */
+
+declare(strict_types=1);
+
+namespace Tests\Core\Common\Domain;
+
+use Core\Common\Domain\LegacyEventEnumTrait;
+
+enum EventEnumStub: string
```
Comment (abuathier) [inline centreon/tests/php/Core/Common/Domain/EventEnumStub.php:28]:
```suggestion
/**
 * This stub exists to test the trait {@see LegacyEventEnumTrait}.
 */
enum EventEnumStub: string
```

File: centreon/src/Core/HostTemplate/Application/UseCase/FindHostTemplates/FindHostTemplates.php (hunk 10 above)
Comment (a-launois) [reply centreon/src/Core/HostTemplate/Application/UseCase/FindHostTemplates/FindHostTemplates.php:114]: This won't work, toBitmask() expect an array not an int

File: centreon/src/Core/HostTemplate/Domain/Model/HostTemplate.php [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+        string $actionUrl = '',
+        ?int $iconId = null,
+        string $iconAlternative = '',
+        string $comment = '',
+        bool $isActivated = true,
+        bool $isLocked = false
+    ) {
+        Assertion::positiveInt($id, 'HostTemplate::id');
+
+        parent::__construct(
```
Comment (callapa) [inline centreon/src/Core/HostTemplate/Domain/Model/HostTemplate.php:122]: Only the mandatory properties must be passed to the constructor.

File: centreon/src/Core/HostTemplate/Domain/Model/HostTemplate.php [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+
+namespace Core\HostTemplate\Domain\Model;
+
+use Assert\AssertionFailedException;
+use Centreon\Domain\Common\Assertion\Assertion;
+use Core\Common\Domain\HostEvent;
+use Core\Common\Domain\SnmpVersion;
+use Core\Common\Domain\YesNoDefault;
+
+class HostTemplate extends NewHostTemplate
```
Comment (callapa) [inline centreon/src/Core/HostTemplate/Domain/Model/HostTemplate.php:32]: You must reverse the dependency, the NewHostTemple class must inherit from the HostTemplate class.

File: centreon/src/Core/Common/Domain/YesNoDefault.php [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * This enum is to be used to handle properties with only three accepted value : Yes, No and Default.
+ */
+enum YesNoDefault:string
+{
+    case No = '0';
+    case Yes = '1';
+    case Default = '2';
+
+	/**
```
Comment (abuathier) [inline centreon/src/Core/Common/Domain/YesNoDefault.php:35]: Be careful : tabs `\t` insteadof spaces `' '`

File: centreon/src/Core/HostTemplate/Application/UseCase/FindHostTemplates/FindHostTemplates.php [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+
+    /**
+     * @param FindHostTemplatesPresenterOnPrem|FindHostTemplatesPresenterSaas $presenter
+     */
+    public function __invoke(PresenterInterface $presenter): void
+    {
+        try {
+            if (
+                ! $this->user->hasTopologyRole(Contact::ROLE_CONFIGURATION_HOSTS_TEMPLATES_READ)
+                && ! $this->user->hasTopologyRole(Contact::ROLE_CONFIGURATION_HOSTS_TEMPLATES_READ_WRITE)
```
Comment (a-launois) [inline centreon/src/Core/HostTemplate/Application/UseCase/FindHostTemplates/FindHostTemplates.php:59]: remove read write

File: centreon/src/Core/HostTemplate/Infrastructure/API/FindHostTemplates/FindHostTemplatesPresenterOnPrem.php [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+                'max_check_attempts' => $hostTemplate['maxCheckAttempts'],
+                'normal_check_interval' => $hostTemplate['normalCheckInterval'],
+                'retry_check_interval' => $hostTemplate['retryCheckInterval'],
+                'is_active_check_enabled' => $hostTemplate['isActiveCheckEnabled'],
+                'isPassive_check_enabled' => $hostTemplate['isPassiveCheckEnabled'],
+                'is_notification_enabled' => $hostTemplate['isNotificationEnabled'],
+                'notification_options' => $hostTemplate['notificationOptions'],
+                'notification_interval' => $hostTemplate['notificationInterval'],
+                'notification_timeperiod_id' => $hostTemplate['notificationTimeperiodId'],
+                // TODO : should the following two conditioned by options.inheritance_mode === '1' ?
```
Comment (abuathier) [inline centreon/src/Core/HostTemplate/Infrastructure/API/FindHostTemplates/FindHostTemplatesPresenterOnPrem.php:71]: TODO to remove

Comment (a-launois) [inline centreon/src/Core/HostTemplate/Infrastructure/API/FindHostTemplates/FindHostTemplatesPresenterOnPrem.php:71]: remove todo

File: centreon/src/Core/HostTemplate/Application/UseCase/FindHostTemplates/FindHostTemplates.php [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+                'snmpCommunity' => $hostTemplate->getSnmpCommunity(),
+                'timezoneId' => $hostTemplate->getTimezoneId(),
+                'severityId' => $hostTemplate->getSeverityId(),
+                'checkCommandId' => $hostTemplate->getCheckCommandId(),
+                'checkCommandArgs' => $hostTemplate->getCheckCommandArgs(),
+                'checkTimeperiodId' => $hostTemplate->getCheckTimeperiodId(),
+                'maxCheckAttempts' => $hostTemplate->getMaxCheckAttempts(),
+                'normalCheckInterval' => $hostTemplate->getNormalCheckInterval(),
+                'retryCheckInterval' => $hostTemplate->getretryCheckInterval(),
+                'isActiveCheckEnabled' => $hostTemplate->isActiveCheckEnabled()->toInt(),
```
Comment (callapa) [inline centreon/src/Core/HostTemplate/Application/UseCase/FindHostTemplates/FindHostTemplates.php:104]: Boolean property must return a boolean value. In this case, yes, false or null but not a integer

**Review by abuathier — DISMISSED** [review]

File: centreon/src/Core/HostTemplate/Domain/Model/NewHostTemplate.php [hunk 24]
```diff
@@ -0,0 +1,387 @@
...
+        $this->snmpCommunity = trim($snmpCommunity);
+        $this->checkCommandArgs = trim($checkCommandArgs);
+        $this->eventHandlerCommandArgs = trim($eventHandlerCommandArgs);
+        $this->noteUrl = trim($noteUrl);
+        $this->note = trim($note);
+        $this->actionUrl = trim($actionUrl);
+        $this->iconAlternative = trim($iconAlternative);
+        $this->comment = trim($comment);
+
+        Assertion::notEmptyString($name, "{$shortName}::name");
```
Comment (callapa) [inline centreon/src/Core/HostTemplate/Domain/Model/NewHostTemplate.php:143]: The trim function was applied to the protected property, not to the "name" argument.

File: centreon/tests/php/Core/HostTemplate/Domain/Model/NewHostTemplateTest.php [hunk 25]
```diff
@@ -0,0 +1,282 @@
...
+// mandatory fields
+foreach (
+    [
+        'name',
+        'alias',
+    ] as $field
+) {
+    it(
+        "should throw an exception when host template {$field} is an empty string",
+        fn() => ($this->createHostTemplate)([$field => ''])
```
Comment (callapa) [inline centreon/tests/php/Core/HostTemplate/Domain/Model/NewHostTemplateTest.php:181]: To test the trim, you should send a string containing only empty characters.

**Review by callapa — APPROVED** [review]

**Review by abuathier — APPROVED** [review]

File: centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 26]
```diff
@@ -17561,3 +17561,9 @@ msgstr "Erreur lors de la recherche de la catégorie d'hôtes #%d"
 
 msgid "Error while retrieving a host category"
 msgstr "Erreur lors de la récupération de la catégorie d'hôtes"
+
+msgid "Error while searching for host templates"
+msgstr "Erreur lors de la recherche des templates d'hôtes"
```
Comment (smau-centreon) [inline centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:17566]:
```suggestion
msgstr "Erreur lors de la recherche des modèles d'hôtes"
```

File: centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 27]
```diff
@@ -17561,3 +17561,9 @@ msgstr "Erreur lors de la recherche de la catégorie d'hôtes #%d"
...
+
+msgid "You are not allowed to access host templates"
+msgstr "Vous n'êtes pas autorisé à accéder aux templates d'hôtes"
```
Comment (smau-centreon) [inline centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:17569]:
```suggestion
msgstr "Vous n'êtes pas autorisé à accéder aux modèles d'hôtes"
```

**Review by smau-centreon — CHANGES_REQUESTED** [review]

**Review by smau-centreon — DISMISSED** [review]

**Review by abuathier — DISMISSED** [review]

**Review by callapa — DISMISSED** [review]

File: centreon/doc/API/centreon-api-v23.10.yaml [hunk 28]
```diff
@@ -6587,6 +6640,23 @@ components:
             With the default time unit of 60s, this number will mean multiples of 1 minute.
 
             A value of 0 disables re-notifications of contacts about problems for this host - only one problem notification will be sent out.
+        notification_timeperiod_id:
+          type: integer
+          description: "Notification tiemperiod ID"
```
Comment (callapa) [inline centreon/doc/API/centreon-api-v23.10.yaml:6645]:
```suggestion
          description: "Notification timeperiod ID"
```

File: centreon/doc/API/centreon-api-v23.10.yaml [hunk 29]
```diff
@@ -6601,80 +6671,89 @@ components:
...
           nullable: true
           maxLength: 65535
           description: "Define an optional URL that can be used to provide more actions to be performed on the host."
-        notes:
+        icon_id:
+          type: integer
+          nullable: true
+          description: "Define the image ID that should be associated with this host template"
+          example: 1
+        alternative_icon:
```
Comment (a-launois) [inline centreon/doc/API/centreon-api-v23.10.yaml:6744]:
```suggestion
        icon_alternative:
```

**Review by abuathier — APPROVED** [review]

**Review by smau-centreon — APPROVED** [review]

**Review by callapa — APPROVED** [review]

**Pull request merged**